File: backend_java/plm/src/main/java/com/plm/plm/Controllers/ProductController.java
package com.plm.plm.Controllers;

import com.plm.plm.dto.ApiResponseDTO;
import com.plm.plm.dto.BOMDTO;
import com.plm.plm.dto.BOMItemDTO;
import com.plm.plm.dto.ProductDTO;
import com.plm.plm.security.JwtTokenProvider;
import com.plm.plm.services.ProductService;
import jakarta.servlet.http.HttpServletRequest;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/products")
public class ProductController {

    @Autowired
    private ProductService productService;

    @Autowired
    private JwtTokenProvider jwtTokenProvider;

    @PostMapping
    public ResponseEntity<ApiResponseDTO<Map<String, ProductDTO>>> createProduct(
            @RequestBody ProductDTO productDTO) {
        ProductDTO product = productService.createProduct(productDTO);
        Map<String, ProductDTO> data = new HashMap<>();
        data.put("product", product);
        return ResponseEntity.status(HttpStatus.CREATED)
            .body(ApiResponseDTO.success(data));
    }

    @GetMapping
    public ResponseEntity<ApiResponseDTO<Map<String, List<ProductDTO>>>> getAllProducts(
            @RequestParam(required = false) String tipo,
            @RequestParam(required = false) String categoria,
            @RequestParam(required = false) String search) {
        List<ProductDTO> products = productService.getAllProducts(tipo, categoria, search);
        Map<String, List<ProductDTO>> data = new HashMap<>();
        data.put("products", products);
        return ResponseEntity.ok(ApiResponseDTO.success(data));
    }

    @GetMapping("/{id}")
    public ResponseEntity<ApiResponseDTO<Map<String, Object>>> getProductById(@PathVariable Integer id) {
        Map<String, Object> data = productService.getProductById(id);
        return ResponseEntity.ok(ApiResponseDTO.success(data));
    }

    @PutMapping("/{id}")
    public ResponseEntity<ApiResponseDTO<Map<String, ProductDTO>>> updateProduct(
            @PathVariable Integer id,
            @RequestBody ProductDTO productDTO) {
        ProductDTO product = productService.updateProduct(id, productDTO);
        Map<String, ProductDTO> data = new HashMap<>();
        data.put("product", product);
        return ResponseEntity.ok(ApiResponseDTO.success(data));
    }

    @PostMapping("/{id}/bom")
    public ResponseEntity<ApiResponseDTO<Map<String, BOMDTO>>> createOrUpdateBOM(
            @PathVariable Integer id,
            @RequestBody BOMDTO bomDTO,
            HttpServletRequest httpRequest) {
        Integer userId = getUserIdFromRequest(httpRequest);
        BOMDTO bom = productService.createOrUpdateBOM(id, bomDTO, userId);
        Map<String, BOMDTO> data = new HashMap<>();
        data.put("bom", bom);
        return ResponseEntity.ok(ApiResponseDTO.success(data));
    }

    @GetMapping("/{id}/bom/history")
    public ResponseEntity<ApiResponseDTO<Map<String, List<BOMDTO>>>> getBOMHistory(
            @PathVariable Integer id) {
        List<BOMDTO> history = productService.getBOMHistory(id);
        Map<String, List<BOMDTO>> data = new HashMap<>();
        data.put("history", history);
        return ResponseEntity.ok(ApiResponseDTO.success(data));
    }

    @PostMapping("/boms/{bomId}/items")
    public ResponseEntity<ApiResponseDTO<Map<String, BOMItemDTO>>> addMaterialToBOM(
            @PathVariable Integer bomId,
            @RequestBody BOMItemDTO bomItemDTO) {
        BOMItemDTO item = productService.addMaterialToBOM(bomId, bomItemDTO);
        Map<String, BOMItemDTO> data = new HashMap<>();
        data.put("item", item);
        return ResponseEntity.status(HttpStatus.CREATED)
            .body(ApiResponseDTO.success(data));
    }

    @GetMapping("/boms/{bomId}")
    public ResponseEntity<ApiResponseDTO<Map<String, BOMDTO>>> getBOMWithItems(
            @PathVariable Integer bomId) {
        BOMDTO bom = productService.getBOMWithItems(bomId);
        Map<String, BOMDTO> data = new HashMap<>();
        data.put("bom", bom);
        return ResponseEntity.ok(ApiResponseDTO.success(data));
    }

    @PutMapping("/bom-items/{itemId}")
    public ResponseEntity<ApiResponseDTO<Map<String, BOMItemDTO>>> updateBOMItem(
            @PathVariable Integer itemId,
            @RequestBody BOMItemDTO bomItemDTO) {
        BOMItemDTO item = productService.updateBOMItem(itemId, bomItemDTO);
        Map<String, BOMItemDTO> data = new HashMap<>();
        data.put("item", item);
        return ResponseEntity.ok(ApiResponseDTO.success(data));
    }

    @DeleteMapping("/bom-items/{itemId}")
    public ResponseEntity<ApiResponseDTO<Map<String, String>>> deleteBOMItem(@PathVariable Integer itemId) {
        productService.deleteBOMItem(itemId);
        Map<String, String> data = new HashMap<>();
        data.put("message", "Item eliminado correctamente");
        return ResponseEntity.ok(ApiResponseDTO.success(data));
    }

    private Integer getUserIdFromRequest(HttpServletRequest request) {
        String authHeader = request.getHeader("Authorization");
        if (authHeader != null && authHeader.startsWith("Bearer ")) {
            String token = authHeader.substring(7);
            return jwtTokenProvider.getUserIdFromToken(token);
        }
        throw new RuntimeException("Token no encontrado");
    }
}

File: Back_end/src/controllers/ProductController.js
import ProductService from '../services/ProductService.js';

/**
 * Controlador de Productos
 * Controladores delgados: solo orquestan, la lógica está en los servicios
 */
class ProductController {
  constructor() {
    this.productService = ProductService;
  }

  /**
   * Crea un nuevo producto
   * POST /api/products
   */
  createProduct = async (req, res, next) => {
    try {
      const product = await this.productService.createProduct(req.body);
      res.status(201).json({
        success: true,
        data: { product: product.toJSON() }
      });
    } catch (error) {
      next(error);
    }
  };

  /**
   * Obtiene todos los productos
   * GET /api/products
   */
  getAllProducts = async (req, res, next) => {
    try {
      const filters = {
        tipo: req.query.tipo,
        categoria: req.query.categoria,
        search: req.query.search
      };
      const products = await this.productService.getAllProducts(filters);
      res.status(200).json({
        success: true,
        data: { products }
      });
    } catch (error) {
      next(error);
    }
  };

  /**
   * Obtiene un producto por ID con su BOM
   * GET /api/products/:id
   */
  getProductById = async (req, res, next) => {
    try {
      const data = await this.productService.getProductById(req.params.id);
      res.status(200).json({
        success: true,
        data
      });
    } catch (error) {
      next(error);
    }
  };

  /**
   * Actualiza un producto
   * PUT /api/products/:id
   */
  updateProduct = async (req, res, next) => {
    try {
      const product = await this.productService.updateProduct(req.params.id, req.body);
      res.status(200).json({
        success: true,
        data: { product: product.toJSON() }
      });
    } catch (error) {
      next(error);
    }
  };

  /**
   * Crea o actualiza el BOM de un producto
   * POST /api/products/:id/bom
   */
  createOrUpdateBOM = async (req, res, next) => {
    try {
      const bom = await this.productService.createOrUpdateBOM(
        req.params.id,
        req.body,
        req.user.id
      );
      res.status(200).json({
        success: true,
        data: { bom: bom.toJSON() }
      });
    } catch (error) {
      next(error);
    }
  };

  /**
   * Agrega un material al BOM
   * POST /api/products/boms/:bomId/items
   */
  addMaterialToBOM = async (req, res, next) => {
    try {
      const item = await this.productService.addMaterialToBOM(req.params.bomId, req.body);
      res.status(201).json({
        success: true,
        data: { item: item.toJSON() }
      });
    } catch (error) {
      next(error);
    }
  };

  /**
   * Obtiene el BOM completo con sus items
   * GET /api/products/boms/:bomId
   */
  getBOMWithItems = async (req, res, next) => {
    try {
      const bom = await this.productService.getBOMWithItems(req.params.bomId);
      res.status(200).json({
        success: true,
        data: { bom }
      });
    } catch (error) {
      next(error);
    }
  };

  /**
   * Actualiza un item del BOM
   * PUT /api/products/bom-items/:itemId
   */
  updateBOMItem = async (req, res, next) => {
    try {
      const item = await this.productService.updateBOMItem(req.params.itemId, req.body);
      res.status(200).json({
        success: true,
        data: { item: item.toJSON() }
      });
    } catch (error) {
      next(error);
    }
  };

  /**
   * Elimina un item del BOM
   * DELETE /api/products/bom-items/:itemId
   */
  deleteBOMItem = async (req, res, next) => {
    try {
      await this.productService.deleteBOMItem(req.params.itemId);
      res.status(200).json({
        success: true,
        message: 'Item eliminado correctamente'
      });
    } catch (error) {
      next(error);
    }
  };

  /**
   * Obtiene el historial de versiones del BOM
   * GET /api/products/:id/bom/history
   */
  getBOMHistory = async (req, res, next) => {
    try {
      const history = await this.productService.getBOMHistory(req.params.id);
      res.status(200).json({
        success: true,
        data: { history }
      });
    } catch (error) {
      next(error);
    }
  };
}

export default new ProductController();
